lib/iris/analysis/__init__.py

```diff
@@ -1419,36 +1419,45 @@ def interp_order(length):
 """
 
 
-def _build_dask_mdtol_function(dask_nanfunction):
+def _build_dask_mdtol_function(dask_stats_function):
     """
     Make a wrapped dask statistic function that supports the 'mdtol' keyword.
 
-    'dask_nanfunction' must be a statistical function that treats NaNs as
-    missing data, and which has the basic call signature
-    "dask_nanfunction(data, axis, **kwargs)".
+    'dask_function' must be a dask statistical function, compatible with the
+    call signature : "dask_stats_function(data, axis, **kwargs)".
+    It must be masked-data tolerant, i.e. it ignores masked input points and
+    performs a calculation on only the unmasked points.
+    For example, mean([1, --, 2]) = (1 + 2) / 2 = 1.5.
 
     The returned value is a new function operating on dask arrays.
-    It has the call signature "stat(data, axis=-1, mdtol=None, *kwargs)".
+    It has the call signature "stat(data, axis=-1, mdtol=None, **kwargs)".
 
     """
-    @wraps(dask_nanfunction)
+    @wraps(dask_stats_function)
     def inner_stat(array, axis=-1, mdtol=None, **kwargs):
-        dask_result = dask_nanfunction(array, axis=axis, **kwargs)
+        # Call the statistic to get the basic result (missing-data tolerant).
+        dask_result = dask_stats_function(array, axis=axis, **kwargs)
         if mdtol is None:
             result = dask_result
         else:
-            point_counts = np.sum(np.ones(array.shape), axis=axis)
-            point_mask_counts = da.sum(da.isnan(array), axis=axis)
+            # Build a lazy computation to compare the fraction of missing
+            # input points at each output point to the 'mdtol' threshold.
+            point_counts = da.sum(da.ones(array.shape, chunks=array.chunks),
```

Comment:
Because the summation may be over one or multiple dimensions (axes).
E.G. if you have shape=(2,3,4), some alternatives might be ...
```
>>> a = np.ones((2,3,4))
>>> a
array([[[ 1.,  1.,  1.,  1.],
        [ 1.,  1.,  1.,  1.],
        [ 1.,  1.,  1.,  1.]],

       [[ 1.,  1.,  1.,  1.],
        [ 1.,  1.,  1.,  1.],
        [ 1.,  1.,  1.,  1.]]])
>>> np.sum(a, axis=1)
array([[ 3.,  3.,  3.,  3.],
       [ 3.,  3.,  3.,  3.]])
>>> np.sum(a, axis=-1)
array([[ 4.,  4.,  4.],
       [ 4.,  4.,  4.]])
>>> np.sum(a)
24.0
>>> np.sum(a, axis=(1,2))
array([ 12.,  12.])
>>> 
```

There is almost certainly a better way of doing that, but I thought this would be okay + at least it ensures that we treat the 'axis' parameter in the prescribed numpy manner.